src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp

```diff
@@ -17,39 +18,83 @@ namespace RelativisticEuler {
 namespace Valencia {
 
 template <size_t Dim>
-std::array<DataVector, Dim + 2> characteristic_speeds(
+void characteristic_speeds(
+    const gsl::not_null<std::array<DataVector, Dim + 2>*> char_speeds,
     const Scalar<DataVector>& lapse, const tnsr::I<DataVector, Dim>& shift,
     const tnsr::I<DataVector, Dim>& spatial_velocity,
     const Scalar<DataVector>& spatial_velocity_squared,
     const Scalar<DataVector>& sound_speed_squared,
     const tnsr::i<DataVector, Dim>& normal) noexcept {
-  auto characteristic_speeds =
-      make_array<Dim + 2>(DataVector(-1.0 * get(dot_product(normal, shift))));
+  const size_t num_grid_points = get<0>(shift).size();
+  // Allocating a single large buffer is much faster than many small buffers
+  Variables<
+      tmpl::list<Tags::TempScalar<0>, Tags::TempScalar<1>, Tags::TempScalar<2>,
+                 Tags::TempScalar<3>, Tags::TempScalar<4>>>
```

Comment:
If you want to use a Variables here, could you not just define a couple of SimpleTags in this CPP file and use them? For instance, one named `LapseTimesNormalVelocity` instead of `TempScalar<1>. These TempTensor tags are very undescriptive, it‘s like making it OK to use variable names `temp_scalar_1` instead of `lapse_times_normal_velocity` in our code.

Reply:
Better than having `temp_scalar_1` variable names I think 😄 Also writing tags isn't too bad, it's 3 lines per tag (2 if we made `name()` use the type name by default if that's possible). It's only the tag name and its type, so that's the same amount of information as declaring a variable.

Reply:
I'm not saying the variable names shouldn't be descriptive, and there is also nothing wrong with `temp_scalar_1` if something doesn't have any deeper meaning. If you want me to write the tags everywhere I'm going back to unique_ptr because that's less work for me.

Reply:
@nilsleiffischer would it be helpful if the tags were named, for instance, `ScalarBuffer` instead of `TempScalar`? Would this name better convey the idea the tag isn't about the value, but about the memory?